core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,232 @@ class DAGSchedulerSuite
+   * Specifically, stage0 fails twice, and then stage1 twice. In total, the job has had four
+   * failures overall but not four failures for a particular stage, and as such should not be
+   * aborted.
+   */
+  test("Failures in different stages should not trigger an overall abort") {
+    // Create a new Listener to confirm that the listenerBus sees the JobEnd message
+    // when we abort the stage. This message will also be consumed by the EventLoggingListener
+    // so this will propagate up to the user.
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+
+    for (attempt <- 0 to Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      val stage0Attempt = taskSets.last
+
+      // Confirm  that this is the first attempt for stage 0
+      checkStageIdAndPrint(0, attempt, stage0Attempt)
+
+      if (attempt < Stage.MAX_STAGE_FAILURES/2) {
+        // Make each task in stage 0 success
+        val completions = stage0Attempt.tasks.zipWithIndex.map{ case (task, idx) =>
+          (Success, makeMapStatus("host" + ('A' + idx).toChar, 2))
+        }.toSeq
+
+        // Run stage 0
+        complete(stage0Attempt, completions)
+
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // We will have one fetch failure for this task set
+        val stage1Attempt = taskSets.last
+        checkStageIdAndPrint(1, attempt, stage1Attempt)
+
+        val stage1Successes = stage1Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run Stage 1, this time with a task failure
+        complete(stage1Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage1Successes
+        )
+      } else {
+        val stage0Successes = stage0Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run stage 0 and fail
+        complete(stage0Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage0Successes
+        )
+      }
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+
+      // In this test case, we're confirming that we never actually fail out (so we should have a
+      // stage that has been resubmitted)
+      assert(scheduler.runningStages.nonEmpty)
+      assert(!ended)
+    }
```

Comment:
I just want to make sure that we can actually finish the job successfully -- you could imagine your checks all passing, but the job is stuck in some weird state where the DAGSCheduler thinks the stage is still running, but no task sets actually created, or something weird like that.  Your asserts are almost there, but not quite.  the final call to `scheduler.resubmitFailedStages` should trigger another attempt for stage 0, and after completing that, you'd get another attempt for stage 1, and then finally the job should complete, but successfully.

I know this is minor, I'd just like to be extra thorough.